packages/quantic/decisions/0001-testing-strategy.md

```diff
@@ -0,0 +1,57 @@
+# Decision Record: Testing Strategy for Quantic
+
+- **Date**: 2024-11-06
+- **Status**: accepted
+
+## Context
+
+Our project involves testing Lightning Web Components (LWCs) to ensure quality and reliability. We have decided to use Jest for unit testing to verify component logic and isolated behavior and Playwright for end-to-end (E2E) testing to validate user interactions and integration within the full UI. This document outlines the criteria for deciding what should be tested in each environment.
+
+## Decision
+
+We have established the following guidelines for Jest and Playwright testing in LWCs:
+
+### Jest Unit Tests
+
+Jest will focus on isolated component behavior, covering specific scenarios without involving the browser environment or real user workflows.
+
+Key points for Jest unit tests:
+
+- **Isolated Component Behavior**: Tests should verify the component's behavior in isolation, without dependency on external systems or other components.
+- **Mock External Dependencies**: Use mocks to simulate external dependencies, such as capabilities used from the Headless library, ensuring unit tests are fast and isolated.
+- **HTML Rendering**: Validate that the component renders the correct HTML based on its state and input properties.
+- **Property Updates**: Ensure component properties update as expected in response to user interactions and lifecycle events.
+- **Edge Cases**: Cover edge cases to ensure the component handles unexpected or extreme inputs gracefully.
+
+**Example Jest Test Scenarios**:
+
+- Mock a Headless controller call to test a specific component in isolation.
+- Mock a Headless state and ensure the component renders correctly by reflecting the state in the UI.
+- Verify that clicking a button updates a specific property and that the rendered HTML reflects this change.
+- Test a lifecycle event, such as `connectedCallback`, to confirm it initializes the component correctly.
+
+### Playwright E2E Tests
+
+Playwright E2E tests will simulate actual user actions to validate the functionality of LWCs within the full user interface and ensure interaction with external systems.
+
+Key points for Playwright E2E tests:
+
+- **User Workflow Testing**: Focus on the essential paths users take to accomplish their goals with a given component, this for all the different Quantic use cases: Search, Insight, Case assist and Recomendations.
+- **Simulate User Actions**: Test realistic user actions, such as clicking buttons, entering data, and navigating through the UI.
+- **External System Interactions**: Verify interactions with external systems, like Coveo APIs and analytics, to ensure components behave as expected in a real-world environment.
```

Comment:
```suggestion
- **External System Interactions**: Verify interactions with external systems, like Coveo APIs and analytics, to ensure components behave as expected in a real-world environment. This also acts as a double test because it will also notify us by failing the test if the external API contract is broken too because we test for the expected request body but also the expected response as much as possible.
```